### PR TITLE
issue 12: copy the state before dispatching it to subscribers

### PR DESCRIPTION
**Pull request by stephanenicolas**
Solves issue #12.

TDD driven.

### CODE REVIEW

**Comment by yichenwu** [comment]
have been tested in real case and work well in main Groupon app

File: grox-core/src/test/java/com/groupon/grox/StoreTest.java [hunk 1]
```diff
@@ -259,4 +260,49 @@ public void testChainState_shouldReturnDifferentStateBeforeAndAfterChainProceed(
...
+  @Test
+  //TDD for issue #12: https://github.com/groupon/grox/issues/12
+  public void testChainState_() {
+    //GIVEN
+    Store<Integer> store = new Store<>(0);
+    StateChangeListener<Integer> mockListener0 = createMock(StateChangeListener.class);
+    StateChangeListener<Integer> mockListener1 = createMock(StateChangeListener.class);
+
+    mockListener0.onStateChanged(0);
+    expectLastCall();
```
Comment (saguinav) [inline grox-core/src/test/java/com/groupon/grox/StoreTest.java]: This is not needed.

File: grox-core/src/test/java/com/groupon/grox/StoreTest.java [hunk 2]
```diff
@@ -259,4 +260,49 @@ public void testChainState_shouldReturnDifferentStateBeforeAndAfterChainProceed(
...
+
+    mockListener1.onStateChanged(0);
+    expectLastCall();
```
Comment (saguinav) [inline grox-core/src/test/java/com/groupon/grox/StoreTest.java]: This is not needed.

**Comment by saguinav** [comment]
Fixed in #15

**Pull request closed without merging**